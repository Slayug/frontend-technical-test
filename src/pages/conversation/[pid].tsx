import ConversationWrapper from "../../components/conversationWrapper/ConversationWrapper";
import {useRouter} from "next/router";
import {useQuery} from "@tanstack/react-query";
import {fetchConversationById} from "../../api/ConversationApi";
import {Spin} from "antd";
import {Fragment} from "react";

export default function Conversation() {

  const router = useRouter()
  const {pid} = router.query as { pid: string | undefined }

  const {data: conversation, isLoading} = useQuery({
    queryKey: ['conversation', pid],
    queryFn: () => fetchConversationById(parseInt(pid)),
    enabled: pid !== undefined
  })

  return <Fragment>
    {isLoading && <Spin/>}
    {conversation && conversation[0] ? <ConversationWrapper key={conversation[0].id} conversation={conversation[0]}/> :
      <p>Conversation not found</p>}
  </Fragment>
}
