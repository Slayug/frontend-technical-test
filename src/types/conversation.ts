export interface Conversation {
  id: number
  recipientId: number
  recipientNickname: string
  senderId: number
  senderNickname: string
  lastMessageTimestamp: number
}

export type ConversationPostDto = Pick<Conversation, "recipientId" | "senderId">
