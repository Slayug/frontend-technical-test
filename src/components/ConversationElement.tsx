import {Conversation} from "../types/conversation";

export default function ConversationElement({ conversation }: { conversation: Conversation }) {


  return <div>
    { conversation.recipientNickname }
  </div>

}
