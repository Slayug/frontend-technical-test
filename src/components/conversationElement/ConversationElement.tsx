import {Conversation} from "../../types/conversation";

import styles from './ConversationElement.module.scss';
import {Avatar} from "antd";

export default function ConversationElement({conversation}: { conversation: Conversation }) {

  function renderLastMessageTimestamp() {
    return new Date(conversation.lastMessageTimestamp * 1000).toLocaleDateString()
  }


  return <a className={styles.conversationElement}>
    <Avatar
      src={`https://joeschmoe.io/api/v1/${conversation.recipientNickname}`}
      shape="circle"
      style={{width: "4.5em", height: "4.5em", backgroundColor: 'white'}}
    />
    <div className={styles.info}>
      <span className={styles.nickname}>{conversation.recipientNickname}</span>
      <span>{renderLastMessageTimestamp()}</span>
    </div>
  </a>

}
