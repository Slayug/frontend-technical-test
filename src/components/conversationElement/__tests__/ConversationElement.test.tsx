import {render, screen} from "@testing-library/react";
import {conversationElodieThibaut, conversationPatrickThibaut} from "../../../utils/dataTests/Conversation.data";
import ConversationElement from "../ConversationElement";
import renderer from "react-test-renderer";

import '../../../../jest/moduleMapper/matchMediaMock';
import UserContextProvider from "../../../contexts/UserContext";
import {Conversation} from "../../../types/conversation";

/**
 * Current user is Thibaut with first ID
 * @param conversation
 */
function renderWithContext(conversation: Conversation) {
  return <UserContextProvider currentUserId={1}>
    <ConversationElement conversation={conversation}/>
  </UserContextProvider>
}

describe("ConversationElement", () => {
  it("renders without crashing", () => {
    render(renderWithContext(conversationPatrickThibaut));
  });

  test("component snapshot", () => {
    const tree = renderer.create(renderWithContext(conversationElodieThibaut)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correct nickname when remote is recipient', () => {
    render(renderWithContext(conversationPatrickThibaut));
    expect(
      screen.getByText(conversationPatrickThibaut.recipientNickname)
    ).toBeInTheDocument()

  });

  it('should render correct nickname when remote is sender', () => {
    render(renderWithContext(conversationElodieThibaut));
    expect(
      screen.getByText(conversationElodieThibaut.senderNickname)
    ).toBeInTheDocument()

  });
});
