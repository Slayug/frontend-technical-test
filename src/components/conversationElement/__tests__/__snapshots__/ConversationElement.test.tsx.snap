// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationElement component snapshot 1`] = `
<a
  className="conversationElement"
>
  <span
    className="ant-avatar ant-avatar-circle ant-avatar-image"
    style={
      {
        "backgroundColor": "white",
        "height": "4.5em",
        "width": "4.5em",
      }
    }
  >
    <img
      onError={[Function]}
      src="https://joeschmoe.io/api/v1/Elodie"
    />
  </span>
  <div
    className="info"
  >
    <span
      className="nickname"
    >
      Elodie
    </span>
    <span>
      07/07/2021
    </span>
  </div>
</a>
`;
